File: frontend/src/types/quote.ts
export interface DiscountConfig {
  saas_year1_pct?: number; // Percentage discount on SaaS Year 1
  saas_all_years_pct?: number; // Percentage discount on SaaS all years
  setup_fixed?: number; // Fixed dollar discount on Setup
  setup_pct?: number; // Percentage discount on Setup
}

export interface ClientData {
  [key: string]: string | number | boolean | null | undefined;
}

export interface TravelConfig {
  [key: string]: string | number | boolean | null | undefined;
}

export interface Quote {
  Id: string;
  QuoteNumber: string;
  ClientName: string;
  ClientOrganization: string | null;
  CreatedBy: string;
  CreatedAt: string;
  UpdatedAt: string;
  Status: string;
}

export interface QuoteVersion {
  Id: string;
  QuoteId: string;
  VersionNumber: number;
  VersionDescription: string | null;
  PricingVersionId: string;
  ClientData: ClientData;
  ProjectionYears: number;
  EscalationModel: string;
  MultiYearFreezeYears: number | null;
  LevelLoadingEnabled: boolean;
  TellerPaymentsEnabled: boolean;
  DiscountConfig: DiscountConfig | null;
  ReferrerId: string | null;
  ReferralRateOverride: number | null;
  MilestoneStyle: string;
  InitialPaymentPercentage: number;
  ProjectDurationMonths: number;
  TravelZoneId: string | null;
  TravelConfig: TravelConfig | null;
  TotalSaaSMonthly: number | null;
  TotalSaaSAnnualYear1: number | null;
  TotalSetupPackages: number | null;
  TotalTravel: number | null;
  TotalContractedAmount: number | null;
  CreatedBy: string;
  CreatedAt: string;
  VersionStatus: string;
  SaaSProducts: QuoteVersionSaaSProduct[];
  SetupPackages: QuoteVersionSetupPackage[];
}

export interface QuoteVersionSaaSProduct {
  Id: string;
  SaaSProductId: string;
  Quantity: number;
  CalculatedMonthlyPrice: number;
  Notes: string | null;
}

export interface QuoteVersionSetupPackage {
  Id: string;
  SKUDefinitionId: string;
  Quantity: number;
  CalculatedPrice: number;
  CustomScopeNotes: string | null;
  SequenceOrder: number | null;
}

export interface QuoteWithVersions extends Quote {
  Versions: QuoteVersion[];
}

export interface NewQuote {
  ClientName: string;
  ClientOrganization: string;
  CreatedBy: string;
}

export interface NewQuoteVersion {
  PricingVersionId: string;
  ClientData: ClientData;
  ProjectionYears: number;
  CreatedBy: string;
  SaaSProducts: {
    SaaSProductId: string;
    Quantity: string;
    Notes?: string;
  }[];
  SetupPackages: {
    SKUDefinitionId: string;
    Quantity: number;
    CustomScopeNotes?: string;
    SequenceOrder?: number;
  }[];
}

export interface PricingVersion {
  Id: string;
  VersionNumber: string;
  Description: string | null;
  IsCurrent: boolean;
}

export interface SaaSProduct {
  Id: string;
  ProductCode: string;
  Name: string;
  Description: string | null;
  Category: string;
  PricingModel: string;
  Tier1Min: number;
  Tier1Max: number;
  Tier1Price: number;
  Tier2Min: number | null;
  Tier2Max: number | null;
  Tier2Price: number | null;
  Tier3Min: number | null;
  Tier3Max: number | null;
  Tier3Price: number | null;
  IsActive: boolean;
}

export interface SKUDefinition {
  Id: string;
  SKUCode: string;
  Name: string;
  Description: string | null;
  Category: string;
  FixedPrice: number | null;
  IsActive: boolean;
}

File: backend/app/schemas/quote.py
"""Pydantic schemas for Quote and QuoteVersion."""

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


# Quote Schemas
class QuoteBase(BaseModel):
    """Base schema for Quote."""

    ClientName: str = Field(..., max_length=255)
    ClientOrganization: str | None = Field(None, max_length=255)


class QuoteCreate(QuoteBase):
    """Schema for creating a new Quote."""

    CreatedBy: str = Field(..., max_length=255)


class QuoteUpdate(BaseModel):
    """Schema for updating a Quote."""

    ClientName: str | None = Field(None, max_length=255)
    ClientOrganization: str | None = Field(None, max_length=255)
    Status: str | None = Field(None, max_length=50)


class QuoteResponse(QuoteBase):
    """Schema for Quote response."""

    Id: UUID
    QuoteNumber: str
    CreatedBy: str
    CreatedAt: datetime
    UpdatedAt: datetime
    Status: str

    class Config:
        """Pydantic config."""

        from_attributes = True


# QuoteVersion Schemas
class QuoteVersionSaaSProductInput(BaseModel):
    """Input schema for SaaS product in quote."""

    SaaSProductId: UUID
    Quantity: Decimal = Field(..., ge=0)
    Notes: str | None = None


class QuoteVersionSetupPackageInput(BaseModel):
    """Input schema for setup package in quote."""

    SKUDefinitionId: UUID
    Quantity: int = Field(default=1, ge=1)
    CustomScopeNotes: str | None = None
    SequenceOrder: int | None = None


class QuoteVersionBase(BaseModel):
    """Base schema for QuoteVersion."""

    VersionDescription: str | None = None
    PricingVersionId: UUID
    ClientData: dict[str, Any]
    ProjectionYears: int = Field(default=5, ge=1, le=10)
    EscalationModel: str = Field(default="STANDARD_4PCT")
    MultiYearFreezeYears: int | None = None
    LevelLoadingEnabled: bool = False
    TellerPaymentsEnabled: bool = False
    DiscountConfig: dict[str, Any] | None = None
    ReferrerId: UUID | None = None
    ReferralRateOverride: Decimal | None = Field(None, ge=0, le=100)
    MilestoneStyle: str = Field(default="FIXED_MONTHLY")
    InitialPaymentPercentage: Decimal = Field(default=Decimal("25.00"), ge=0, le=100)
    ProjectDurationMonths: int = Field(default=10, ge=1)
    TravelZoneId: UUID | None = None
    TravelConfig: dict[str, Any] | None = None


class QuoteVersionCreate(QuoteVersionBase):
    """Schema for creating a new QuoteVersion."""

    QuoteId: UUID
    SaaSProducts: list[QuoteVersionSaaSProductInput] = Field(default_factory=list)
    SetupPackages: list[QuoteVersionSetupPackageInput] = Field(default_factory=list)
    CreatedBy: str = Field(..., max_length=255)


class QuoteVersionUpdate(BaseModel):
    """Schema for updating a QuoteVersion."""

    VersionDescription: str | None = None
    ClientData: dict[str, Any] | None = None
    ProjectionYears: int | None = Field(None, ge=1, le=10)
    EscalationModel: str | None = None
    MultiYearFreezeYears: int | None = None
    LevelLoadingEnabled: bool | None = None
    TellerPaymentsEnabled: bool | None = None
    DiscountConfig: dict[str, Any] | None = None
    ReferrerId: UUID | None = None
    ReferralRateOverride: Decimal | None = Field(None, ge=0, le=100)
    MilestoneStyle: str | None = None
    InitialPaymentPercentage: Decimal | None = Field(None, ge=0, le=100)
    ProjectDurationMonths: int | None = Field(None, ge=1)
    TravelZoneId: UUID | None = None
    TravelConfig: dict[str, Any] | None = None
    SaaSProducts: list[QuoteVersionSaaSProductInput] | None = None
    SetupPackages: list[QuoteVersionSetupPackageInput] | None = None
    VersionStatus: str | None = None


class QuoteVersionSaaSProductResponse(BaseModel):
    """Response schema for SaaS product in quote."""

    Id: UUID
    SaaSProductId: UUID
    Quantity: Decimal
    CalculatedMonthlyPrice: Decimal
    Notes: str | None

    class Config:
        """Pydantic config."""

        from_attributes = True


class QuoteVersionSetupPackageResponse(BaseModel):
    """Response schema for setup package in quote."""

    Id: UUID
    SKUDefinitionId: UUID
    Quantity: int
    CalculatedPrice: Decimal
    CustomScopeNotes: str | None
    SequenceOrder: int | None

    class Config:
        """Pydantic config."""

        from_attributes = True


class QuoteVersionResponse(QuoteVersionBase):
    """Schema for QuoteVersion response."""

    Id: UUID
    QuoteId: UUID
    VersionNumber: int
    TotalSaaSMonthly: Decimal | None
    TotalSaaSAnnualYear1: Decimal | None
    TotalSetupPackages: Decimal | None
    TotalTravel: Decimal | None
    TotalContractedAmount: Decimal | None
    CreatedBy: str
    CreatedAt: datetime
    VersionStatus: str
    SaaSProducts: list[QuoteVersionSaaSProductResponse] = Field(default_factory=list)
    SetupPackages: list[QuoteVersionSetupPackageResponse] = Field(default_factory=list)

    class Config:
        """Pydantic config."""

        from_attributes = True


class QuoteWithVersionsResponse(QuoteResponse):
    """Schema for Quote with all its versions."""

    versions: list[QuoteVersionResponse] = Field(default_factory=list)

    class Config:
        """Pydantic config."""

        from_attributes = True
